### DISCUSSION TITLE
Victory Points for each Generation

### DISCUSSION THREAD

**Discussion by LordPhobos** [bafolts/terraforming-mars#3535] [Ideas]
Hi.

A friend o'mine suggested this nice feature: would it be possibile to have, at the end of the game, a statistics panel showing the number of vitory points made by each player in each generation, along with simple graphs?

It would be a really nice feature!

Thanks

**Comment by bafolts** [comment]
I _think_ we have this data available while switching to the end of the game. We would have to load all the old game states and calculate this value.

**Comment by LordPhobos** [comment]
Would that be possibile? Can you give it a try?

**Reply by bafolts** [reply]
I don't think we are missing any data here. I can do a small POC this week to see how the solution would look with our current database structure.

**Comment by bafolts** [comment]
During most game actions we store a serialized game object. When transitioning to the end of the game we have these in database. They are deleted during the transition. If we queried all of these objects and deserialized them we could attempt to calculate this value.

I would think the algorithm is something like:

1. Query every saved game instance ordered by save_id
2. Find the last row from every generation.
3. Deserialize these instances and calculate victory points for every player.
4. Store these results in game_results table.

For step 2 we have to deserialize to know which generation the object is from. This operation is going to block the event loop. Without refactoring the database structure I would be concerned that we blocked the event loop. We would have to make sure to calculate this value in chunks. I can't imagine that many games are completing at once on any game server so as long as this operation doesn't block the event loop I assume it would finish within a second.

To allow the event loop to work on other actions I would work on deserializing the rows in small chunks.

I would say this idea is PLAUSIBLE.

**Comment by LordPhobos** [comment]
Great! Looking forward to see how it develop.

**Comment by LordPhobos** [comment]
Still looking forward for that.
Thanks for all the efforts you take for this game.

**Comment by kberg** [comment]
This is now done.
